### ISSUE TITLE
一些pretraining的問題

### ISSUE THREAD

**Issue by ccchang0111** [bojone/bert4keras#40]
非常謝謝苏神這麼用心把bert轉成淺顯易懂的keras版本！

我有3個基本的問題：
1。在`pretraining.py`裡，
https://github.com/bojone/bert4keras/blob/06a9410d2bb2fa26e07c4f0cd544c01919232533/pretraining/roberta/pretraining.py#L70-L87
`mlm_loss` and `mlm_acc` 是被放在train_model裡的最後一層，並不是像一般放在compile的loss裡。如果是放在model裡，目前的mlm_loss和mlm_acc 是return 一個數值，它的shape = (). 這樣會不會有問題？感覺應該要是shape = (None, ). 是不是得在sum裡加個 `axis = 1`？
`loss = K.sum(loss * is_masked, axis=1) / (K.sum(is_masked, axis=1) + K.epsilon() )`

2。在pretraining裡的optimization過程程中是同時optimize loss和acc嗎？
https://github.com/bojone/bert4keras/blob/06a9410d2bb2fa26e07c4f0cd544c01919232533/pretraining/roberta/pretraining.py#L111-L114
我不是很熟悉上面這樣的用法，想了解為什麼不用loss就好，還是我誤會這個用法了？

3。在`data_utils.py`裡，
https://github.com/bojone/bert4keras/blob/06a9410d2bb2fa26e07c4f0cd544c01919232533/pretraining/roberta/data_utils.py#L189-L192
`'mlm_acc': K.zeros([1])` 是不是要改成 `'mlm_acc': K.ones([1])` 因為accuracy max = 1?

再次感謝！

**Comment by bojone** [comment]
keras的标准loss只支持y_true, y_pred输入，但很显然我这里的loss还有is_masked作为输入，标准的loss做不到这一点。另外一种加loss的方式是model.add_loss方法，这种方法比较灵活，但我试过，在tpu上没法跑通，所以只能将它封装为层。

至于你最后说的y的具体值不重要，因为最终外部传入的y都没有用过。

参考：https://kexue.fm/archives/4493

**Comment by ccchang0111** [comment]
謝謝這麼迅速回覆！ 
我看train_model.summary() 最後一層的shape= (). 在compile的時候好像會出問題？

**Comment by bojone** [comment]
我跑过很多次了，没出问题。。。。

**Comment by ccchang0111** [comment]
Haha相信的 謝謝確認！
